src/shell/mod.rs

```diff
@@ -424,20 +453,30 @@ async fn execute_command(
       Ok(value) => value,
       Err(value) => return value,
     };
-
-    match redirect.maybe_fd {
-      Some(RedirectFd::Fd(2)) => (stdout, pipe),
-      Some(RedirectFd::Fd(1)) | None => (pipe, stderr),
-      Some(RedirectFd::Fd(_)) => {
-        let _ = stderr.write_line(
-          "only redirecting to stdout (1) and stderr (2) is supported",
-        );
-        return ExecuteResult::from_exit_code(1);
-      }
-      Some(RedirectFd::StdoutStderr) => (pipe.clone(), pipe),
+    match pipe {
+      RedirectPipe::Input(pipe) => match redirect.maybe_fd {
+        Some(_) => {
+          let _ = stderr.write_line(
+            "input redirects with file descriptors are not supported",
+          );
+          return ExecuteResult::from_exit_code(1);
+        }
+        None => (pipe, stdout, stderr),
+      },
+      RedirectPipe::Output(pipe) => match redirect.maybe_fd {
+        Some(RedirectFd::Fd(2)) => (stdin, stdout, pipe),
```

Comment:
I don't understand this bit - why is there `Fd(2)` and then `RedirectFd::StdoutStderr`?

Reply:
`RedirectFd::StdoutStderr` is to redirect both stdout and stderr.